File: LeetCode/Medium/Design Underground System/Go/design_underground_system.go
package main

type User struct {
    check_in string
    checkin_time int
}

type Pair struct {
    total int
    n int
}

type UndergroundSystem struct {
    user_data map[int]User
    src_des map[string]map[string]Pair
}


func Constructor() UndergroundSystem {
    return UndergroundSystem{user_data:make(map[int]User), src_des:make(map[string]map[string]Pair)}
}


func (this *UndergroundSystem) CheckIn(id int, stationName string, t int)  {
    this.user_data[id] = User{check_in : stationName, checkin_time:t}
}


func (this *UndergroundSystem) CheckOut(id int, stationName string, t int)  {
    time_diff := t-this.user_data[id].checkin_time
    if _,ok:=this.src_des[this.user_data[id].check_in];!ok{
        this.src_des[this.user_data[id].check_in] = make(map[string]Pair)
    }
    val := this.src_des[this.user_data[id].check_in][stationName].total
    n := this.src_des[this.user_data[id].check_in][stationName].n
    this.src_des[this.user_data[id].check_in][stationName] = Pair{total:val+time_diff, n:n+1}
}


func (this *UndergroundSystem) GetAverageTime(startStation string, endStation string) float64 {
    return float64(this.src_des[startStation][endStation].total)/float64(this.src_des[startStation][endStation].n)
}

func main () { }

File: LeetCode/Medium/Design Underground System/Java/designUndergroundSystem.java
class Pair {
    String stationName;
    int t;
    public Pair(String name, int t) {
        this.t = t;
        this.stationName = name;
    }
}

class UndergroundSystem {
    Map<Integer, Pair> checkInTrack;
    Map<String, HashMap<String, Integer>> travel;
    Map<String, HashMap<String, Integer>> times;
    public UndergroundSystem() {
        checkInTrack = new HashMap<>();
        travel = new HashMap<>();
        times = new HashMap<>();
    }
    
    public void checkIn(int id, String stationName, int t) {
        checkInTrack.put(id, new Pair(stationName, t));
    }
    
    public void checkOut(int id, String stationName, int t) {
        Pair details = checkInTrack.get(id);
        int diff = t-details.t;
        travel.computeIfAbsent(details.stationName, k -> new HashMap<>());
        times.computeIfAbsent(details.stationName, k -> new HashMap<>());
        HashMap<String, Integer> toStation = travel.get(details.stationName);
        HashMap<String, Integer> toStationtimes = times.get(details.stationName);
        toStation.put(stationName, toStation.getOrDefault(stationName, 0) + diff);
        toStationtimes.put(stationName, toStationtimes.getOrDefault(stationName, 0)+1);
    }
    
    public double getAverageTime(String startStation, String endStation) {
        return (double)travel.get(startStation).get(endStation)/times.get(startStation).get(endStation);
    }
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem obj = new UndergroundSystem();
 * obj.checkIn(id,stationName,t);
 * obj.checkOut(id,stationName,t);
 * double param_3 = obj.getAverageTime(startStation,endStation);
 */
